jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java

```diff
@@ -209,38 +215,151 @@ private static StringBuilder preprocessLayout(MemoryLayout targetLayout, boolean
+	private static StringBuilder encodeUnionLayoutStr(UnionLayout unionLayout, StringBuilder targetLayoutStr, boolean isDownCall) {
+		long unionByteSize = unionLayout.byteSize();
+		StringBuilder elementLayoutStrs = new StringBuilder("");
+		int byteAlignment = (int)unionLayout.byteAlignment();
+		ValueLayout sequenceElement = null;
+
+		/* The idea is to select the primitve type with the biggest alignement of the union
+		 * (obtained via unionLayout.byteAlignment()) to be laid out as a sequence which is
+		 * correctly mappped onto the given GPRs/FPRs on the hardware.
+		 */
+		switch (byteAlignment) {
+		case 1: /* 1 Byte. */
+			sequenceElement = JAVA_BYTE;
+			break;
+		case 2: /* 2 Bytes. */
+			sequenceElement = JAVA_SHORT;
+			break;
+		case 4: /* 4 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_INT;
+			} else {
+				sequenceElement = JAVA_FLOAT;
 			}
-			targetLayoutString.append('[').append(elementLayoutStrs).append(']');
+			break;
+		case 8: /* 8 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_LONG;
+			} else {
+				sequenceElement = JAVA_DOUBLE;
+			}
+			break;
+		default:
+			throw new IllegalArgumentException("Invalid alignment of the union layout: " + unionLayout);
 		}
 
-		return targetLayoutString;
+		long elementByteSize = sequenceElement.byteSize();
+		long sequenceCount = (0 == (unionByteSize % elementByteSize)) ?
+				(unionByteSize / elementByteSize) : (unionByteSize / elementByteSize + 1);
+
+		/* There is no need to create a sequence layout for a single element. */
+		MemoryLayout sequenceLayout = (sequenceCount > 1) ?
+				MemoryLayout.sequenceLayout(sequenceCount, sequenceElement) : sequenceElement;
+		elementLayoutStrs.append(preprocessLayout(sequenceLayout, isDownCall));
+
+		/* Prefix "#" to denote the start of this layout string in the case of downcall. */
+		if (isDownCall) {
+			targetLayoutStr.append('#').append(1);
+		}
+		targetLayoutStr.append('[').append(elementLayoutStrs).append(']');
+
+		return targetLayoutStr;
 	}
 
+	/* Check whether the specified 4/8-byte aligned group layout (struct/union) layout contains
+	 * non-floating-point primitves so as to determine whether the elements of union can be safely
+	 * mapped to GPRs.
+	 */
+	private static boolean validatePrimTypesOfUnionForGPR(GroupLayout groupLayout, int byteAlignment) {
+		List<MemoryLayout> elementLayoutList = groupLayout.memberLayouts();
+		boolean arePrimTypesForGPR = false;
+		for (MemoryLayout elementLayout : elementLayoutList) {
+			if (((elementLayout instanceof ValueLayout elemValueLayout) && (validatePrimTypeForGPR(elemValueLayout, byteAlignment)))
+				|| ((elementLayout instanceof GroupLayout elemGroupLayout) && validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))
+			) {
+				arePrimTypesForGPR = true;
+				break;
+			}
+		}
```

Comment:
I think this is inverted (all members must be valid, not just one):
```java
		boolean arePrimTypesForGPR = true;
		for (MemoryLayout elementLayout : elementLayoutList) {
			if ((elementLayout instanceof ValueLayout elemValueLayout)
					&& validatePrimTypeForGPR(elemValueLayout, byteAlignment)) {
				// valid, keep checking
			} else if ((elementLayout instanceof GroupLayout elemGroupLayout)
					&& validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment)) {
				// valid, keep checking
			} else {
				arePrimTypesForGPR = false;
				break;
			}
		}
```

Reply:
Not really. The idea is it is always safe map them to the GPRs as long as one element (layout) of the struct/union is validated by `validatePrimTypeForGPR` regardless of the types of other layouts in the struct/union, according to the mapping conventions specific to platforms.

Reply:
Please explain how this makes sense. If a union contains a member that couldn't possibly be passed in a register, how could a compiler pass the union in a register?

Reply:
In most cases/platforms (including exceptions listed in code), a union is simulated by a struct which will be mapped to GPRs as long as any element of the union is detected as a non-floating point primitive, in which case the union will be safely mapped to GPRs with the biggest size & alignment of the union element.

Let's get stared with a couple of simple examples on X86_64 to see how it works to simulate a union with a struct as follows:
typedef int64_t LONG;

1.  `union {float elem1; int elem2; }`:
It is treated as a `struct {int elem;}` (size & alignment are both 4 bytes) which is directly mapped to a GPR.

2.  `union {LONG elem1; double elem2; }`:
It is treated as a `struct {LONG elem; }` (size & alignment are both 8 bytes) which is directly mapped to a GPR.
 
3.  `union {LONG elem1; double elem2; struct { LONG elem1; LONG elem2; } }` (with a nested 16-bytes struct in the union):
It is treated as a `struct {LONG elem1; LONG elem2; }` (size = 16 bytes & alignment = 8 bytes) which is directly mapped to two GPRs (the same as how a normal struct is mapped to GPRs).

Reply:
Your response seems to support my claim: all members must be (directly or indirectly) mappable to registers.

Reply:
Not really.  I previously listed the cases of members of union that fit into registers available for use.

If a member of union (which is actually simulated by a struct) can't fit into the registers (which is likely treated as if as a struct with only one element for this member), the compiler knows exactly how to place them into the overflow/reserved area (based on the ABI documents on given platforms) in the same way as how they deal with a struct with members greater than what the registers can hold in size in such case. That being said, this is no special case from our code perspective given the compiler knows everything as to how to handle such kind of struct (actually a sequence of fixed primitives).

Reply:
Consider
```c
union U64 {
    char   m1[64];
    double m2;
    long   m3;
};
```
On Linux x64, gcc passes `union U64` objects on the stack, contrary to what this would do: The presence of a `long` or `double` members is not sufficient to determine that registers will be used. I repeat my claim that this method should return true if and only if _all_ members have types that would use registers.